module/zfs/zfs_sysfs.c

```diff
@@ -0,0 +1,441 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+/*
+ * Copyright (c) 2018 by Delphix. All rights reserved.
+ */
+
+#include <sys/types.h>
+#include <sys/param.h>
+#include <sys/zfeature.h>
+#include <sys/zfs_sysfs.h>
+#include <sys/kmem.h>
+#include <sys/fs/zfs.h>
+#include <linux/kobject.h>
+
+#include "zfs_prop.h"
+
+#if !defined(_KERNEL)
+#error kernel builds only
+#endif
+
+/*
+ * ZFS Module sysfs support
+ *
+ * This extends our sysfs '/sys/module/zfs' entry to include feature
+ * and property attributes. The primary consumer of this information
+ * is user processes, like the zfs CLI, that need to know what the
+ * current loaded ZFS module supports. The libzfs binary will consult
+ * this information when instantiating the zfs|zpool property tables
+ * and the pool features table.
+ *
+ * The added subdirectories, listed below, contain attributes (files)
+ * /sys/module/zfs
+ *		├── features.pool
+ *		├── features.runtime
+ *		├── properties.dataset
+ *		└── properties.pool
+ *
+ * The local interface for the zfs kobjects includes:
+ *	zfs_kobj_init()
+ *	zfs_kobj_add()
+ *	zfs_kobj_release()
+ *	zfs_kobj_add_attr()
+ *	zfs_kobj_fini()
+ */
+
+/*
+ * ZFS runtime features for '/sys/module/zfs/features.runtime'
+ *
+ * This list is for runtime features that don't have a pool feature
+ * already associated with it.  It should include runtime features
+ * added in ZFS release 0.8.0 and beyond.
```

Comment:
I think this will tend to be a case-by-case decision. I can update the comment to reflect this.

Reply:
@behlendorf  Would it make more sense to make this `features.interface` and as you suggest enumerate features that have a `libzfs` interface exposure? (i.e. additions to `zfs_ioc_t`).
For example,
```
channel_program
encryption_key
device_removal_remap
pool_checkpoint
```

Reply:
That makes the most sense to me if the problem we're trying to solve is to provide a generic mechanism for the kernel to advertise what functionality it supports.  In which case, it seems like the right thing to do would be to generate this list using all the registered ioctls, i.e. `zfs_ioc_vec`.  Just like you've done with the properties and features.  It might also be useful to identify, perhaps in the files contents, if it's a legacy ioctl or new style.